### DISCUSSION TITLE
Reusing PBF Build Data

### DISCUSSION THREAD

**Discussion by pipo3944** [valhalla/valhalla#4749] [Q&A]
Hello,  

I am working on a project that requires running isochrone analysis using OSM data for all of Japan along with GTFS data.

Our GTFS data updates frequently, but the current build process takes over an hour.
Here are the details:

- **Data**:
    - **OSM**: japan-latest, 1.86GB
    - **GTFS**: 28KB (zipped file)
- **Build Time**: 01:15:17

Since the PBF data does not change, I am looking for a way to rebuild only the GTFS data to reduce the build and startup time.

Are there any effective approaches to achieve this?

Thank you for your assistance.

**Comment by nilsnolde** [comment]
sure, just do `valhalla_build_tiles -h`  and inspect the "stages". if you end a tile before transit connects, you can copy & use that temporary artifact for future transit builds.

**Comment by pipo3944** [comment]
@nilsnolde 
Thank you for your kind response!

Following your suggestion, I have successfully built the GTFS data. The isochrone analysis result using the GTFS file was satisfactory for me.
![スクリーンショット 2024-06-11 15 24 21](https://github.com/valhalla/valhalla/assets/12875256/17c302f9-2d17-429c-8aa2-16fba0919068)

However, when I deleted all the GTFS data to rebuild it again, I encountered the following error:

```
$ valhalla_build_tiles -c valhalla.json japan-latest.osm.pbf -s transit
2024/06/11 06:27:39.305811 [INFO] Running valhalla_build_tiles with 2 thread(s).
2024/06/11 06:27:39.305848 [INFO] Start stage = transit End stage = cleanup
2024/06/11 06:27:39.310963 [INFO] Adding 2 transit tiles to the local graph...
2024/06/11 06:27:39.731597 [WARN] Could not create transit connect edge from end node because its in another tile
2024/06/11 06:27:39.788412 [WARN] Could not create transit connect edge from end node because its in another tile
2024/06/11 06:27:39.873217 [WARN] Could not create transit connect edge from end node because its in another tile
2024/06/11 06:27:47.551363 [INFO] Found 578 connection edges
2024/06/11 06:27:47.551391 [INFO] Finished - TransitBuilder took 8 secs
2024/06/11 06:27:47.551951 [INFO] HierarchyBuilder
terminate called after throwing an instance of 'std::runtime_error'
  what():  /usr/local/src/valhalla/valhalla/baldr/graphtile.h:164 GraphTile NodeInfo index out of bounds: 685237,2,0 nodecount= 0
Aborted (core dumped)
```

Do you have any insights or suggestions regarding this error?

Here's what I have done:

①First, I inspected the stages using `valhalla_build_tiles -h`, and here is the output:
```
Build stage strings (in order)
    initialize
    parseways
    parserelations
    parsenodes
    constructedges
    build
    enhance
    filter
    transit
    bss
    hierarchy
    shortcuts
    restrictions
    elevation
    validate
    cleanup
```

I assume that

1. initialize ~ filter: GTFS not included 
2. transit ~ cleanup: GTFS included in build process

②Thus, I attempted to rebuild just the GTFS data by running:
```
valhalla_build_tiles -c valhalla.json japan-latest.osm.pbf -s initialize -e filter
```

③And then, 
```
valhalla_build_tiles -c valhalla.json japan-latest.osm.pbf -s transit
```

It worked by far. 

④After that, I deleted the GTFS data from the containing directory and rebuild it using the following command: 
```
valhalla_build_tiles -c valhalla.json japan-latest.osm.pbf -s transit
```

I got error as shown above. 

Thank you for your assistance.

valhalla_build_tiles 3.4.0

**Comment by chrstnbwnkl** [comment]
You can't rerun parts of the build process on a fully built graph. Try saving your graph tiles after running

```
valhalla_build_tiles -c valhalla.json japan-latest.osm.pbf -s initialize -e filter
```

and then keep re-using that graph as the base graph for subsequent transit builds instead of just deleting the transit level. The graph build fails in your case because there are other stages like hierarchy building that don't work on a graph where that stage already ran (someone correct me if I'm wrong).

**Comment by pipo3944** [comment]
@chrstnbwnkl 
Ah, I see, that makes sense now. So, duplicating the build files generated by 
```
valhalla_build_tiles -c valhalla.json japan-latest.osm.pbf -s initialize -e filter
```
and then rebuilding for GTFS each time is the correct approach.

Thank you so much!
With this, we have managed to reduce the build time for GTFS updates to approximately 5 minutes.
